File: Algorithms/src/664. Strange Printer/Code/solution.js
/**
 * @param {string} s - The input string that the strange printer needs to print.
 * @return {number} - The minimum number of turns the printer needs to print the entire string.
 */
var strangePrinter = function (s) {
  const n = s.length; // Get the length of the input string

  // Initialize a 2D DP array with dimensions n x n and fill it with zeros.
  // dp[i][j] will store the minimum number of turns required to print the substring s[i:j+1].
  const dp = Array.from({ length: n }, () => Array(n).fill(0));

  // Iterate over the string in reverse order to fill the DP table
  for (let i = n - 1; i >= 0; i--) {
    dp[i][i] = 1; // Base case: A single character needs 1 turn to print

    // Consider all substrings starting at index i and ending at index j
    for (let j = i + 1; j < n; j++) {
      // Assume we print the character s[j] separately, hence adding 1 turn
      dp[i][j] = dp[i][j - 1] + 1;

      // Check for possible combinations to minimize the turns
      // Iterate over possible partitions within the substring s[i:j+1]
      for (let k = i; k < j; k++) {
        // If s[k] equals s[j], we can merge the turns
        if (s[k] === s[j]) {
          // Update dp[i][j] with the minimum turns between the current value and the value obtained by merging
          dp[i][j] = Math.min(dp[i][j], dp[i][k] + dp[k + 1][j - 1]);
        }
      }
    }
  }

  // The result is stored in dp[0][n-1], which represents the minimum turns required to print the entire string s.
  return dp[0][n - 1];
};

File: Algorithms/src/664. Strange Printer/Code/solution.java
class Solution {
    public int strangePrinter(String s) {
        int n = s.length(); // Get the length of the string
        int[][] dp = new int[n][n]; // Create a 2D array to store the minimum turns needed to print each substring

        // Iterate over the string in reverse order to fill the DP table
        for (int i = n - 1; i >= 0; i--) {
            dp[i][i] = 1; // A single character always requires 1 turn to print

            // Consider all substrings starting from index 'i' to 'j'
            for (int j = i + 1; j < n; j++) {
                // Start with the assumption that printing s[j] requires an additional turn
                dp[i][j] = dp[i][j - 1] + 1;

                // Check for all possible partitions of the substring s[i...j]
                for (int k = i; k < j; k++) {
                    // If the character at index 'k' matches the character at index 'j'
                    if (s.charAt(k) == s.charAt(j)) {
                        // We can potentially minimize the number of turns by printing s[j] along with
                        // the segment that ends at 'k'
                        dp[i][j] = Math.min(dp[i][j], dp[i][k] + dp[k + 1][j - 1]);
                    }
                }
            }
        }

        // The minimum number of turns needed to print the entire string s[0...n-1] is
        // stored in dp[0][n-1]
        return dp[0][n - 1];
    }
}
